### ISSUE TITLE
Status of rate limiting

### ISSUE THREAD

**Issue by kaysond** [ccxt/ccxt#929]
What is the current status of the rate limiting implementation?

The manual says the "rateLimit" property is not used, but I noticed the initRestRateLimiter() function and the token bucket + throttle.js... And it seems that if you set enableRateLimit, fetch2() will wait on throttle()

Also I believe the rate limit for Gemini is incorrect. It's at 1500, but should probably be 1000. The actual limit is 2 per second but recommended is 1 per second (see [here](https://docs.gemini.com/rest-api/#rate-limits))

**Comment by kroitor** [comment]
> What is the current status of the rate limiting implementation?

Works, but experimental.

> The manual says the "rateLimit" property is not used, but I noticed the initRestRateLimiter() function and the token bucket + throttle.js... And it seems that if you set enableRateLimit, fetch2() will wait on throttle()

The manual is slightly outdated. See the [examples](https://github.com/ccxt/ccxt/tree/master/examples/js). We will be thankful for your contributions to the docs and code. More about it here: https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md

> Also I believe the rate limit for Gemini is incorrect. It's at 1500, but should probably be 1000. The actual limit is 2 per second but recommended is 1 per second

That is the safe default, you can set it to whatever you want upon instantiation...

```
const gemini = new ccxt.gemini ({ enableRateLimit: true, rateLimit: 1000 })
```

We will reconsider and update the default value as well.

Let me know if this answers your questions or not.

**Comment by kaysond** [comment]
@kroitor Thanks.
One more totally unrelated question - the base Exchange class initializes empty objects for some stuff:
```
        this.balance    = {}
        this.orderbooks = {}
        this.tickers    = {}
        this.fees       = {}
        this.orders     = {}
        this.trades     = {}
        this.currencies = {}
```
Are these all supposed to be used? Currencies seems to get populated, but orderbooks and tickers dont

**Comment by kroitor** [comment]
> Are these all supposed to be used?

Not yet, you're supposed to used documented features + examples.

**Comment by kaysond** [comment]
Might be nice to have the relevant methods populate those properties automatically.

Thank you for the fast answers
